File: recursividad/cpp/cola.cpp
#include <iostream>
#include <stdexcept>

using namespace std;

class Nodo {
public:
    int dato;
    Nodo* siguiente;

    Nodo(int dato) {
        this->dato = dato;
        this->siguiente = nullptr;
    }
};

class Cola {
private:
    Nodo* frente;
    Nodo* finalDeCola;

public:
    Cola() {
        this->frente = nullptr;
        this->finalDeCola = nullptr;
    }

    void enqueue(int nuevoDato) {
        Nodo* nuevoNodo = new Nodo(nuevoDato);
        if (finalDeCola == nullptr) {
            frente = nuevoNodo;
            finalDeCola = nuevoNodo;
        } else {
            finalDeCola->siguiente = nuevoNodo;
            finalDeCola = nuevoNodo;
        }
    }

    int dequeue() {
        if (estaVacia()) {
            throw runtime_error("Cola vacia");
        }
        int dato = frente->dato;
        Nodo* nodoAEliminar = frente;
        frente = frente->siguiente;
        
        if (frente == nullptr) {
            finalDeCola = nullptr;
        }
        delete nodoAEliminar;
        return dato;
    }

    int peek() {
        if (estaVacia()) {
            throw runtime_error("Cola vacia");
        }
        return frente->dato;
    }

    bool estaVacia() {
        return frente == nullptr && finalDeCola == nullptr;
    }

    void invertirRecursivo() {
        if (estaVacia()) {
            return;
        }

        int dato = dequeue();
        invertirRecursivo();
        enqueue(dato);
    }

    bool buscarElementoRecursivo(Nodo* nodo, int valor) {
        if (nodo == nullptr) {
            return false;
        }
        if (nodo->dato == valor) {
            return true;
        }
        return buscarElementoRecursivo(nodo->siguiente, valor);
    }

    Nodo* getFrente() {
        return frente;
    }

    void mostrar() {
        Nodo* actual = frente;
        cout << "Cola: ";
        while (actual != nullptr) {
            cout << actual->dato << " ";
            actual = actual->siguiente;
        }
        cout << endl;
    }
};

int main() {
    Cola c;
    c.enqueue(10);
    c.enqueue(20);
    c.enqueue(30);

    cout << "Cola original:" << endl;
    c.mostrar();

    c.invertirRecursivo();
    cout << "Cola invertida recursivamente:" << endl;
    c.mostrar();

    cout << "Buscar 20 (recursivo): " << (c.buscarElementoRecursivo(c.getFrente(), 20) ? "true" : "false") << endl;
    cout << "Buscar 50 (recursivo): " << (c.buscarElementoRecursivo(c.getFrente(), 50) ? "true" : "false") << endl;

    return 0;
}

File: recursividad/python/cola.py
class Nodo:
    def __init__(self, dato):
        self.dato = dato
        self.siguiente = None

class Cola:
    def __init__(self):
        self.frente = None
        self.final_de_cola = None

    def enqueue(self, nuevo_dato):
        nuevo_nodo = Nodo(nuevo_dato)
        if self.final_de_cola is None:
            self.frente = nuevo_nodo
            self.final_de_cola = nuevo_nodo
        else:
            self.final_de_cola.siguiente = nuevo_nodo
            self.final_de_cola = nuevo_nodo

    def dequeue(self):
        if self.esta_vacia():
            raise Exception("Cola vacía")
        dato = self.frente.dato
        self.frente = self.frente.siguiente
        if self.frente is None:
            self.final_de_cola = None
        return dato

    def peek(self):
        if self.esta_vacia():
            raise Exception("Cola vacía")
        return self.frente.dato

    def esta_vacia(self):
        return self.frente is None and self.final_de_cola is None

    def invertir_recursivo(self):
        if self.esta_vacia():
            return
        
        dato = self.dequeue()
        self.invertir_recursivo()
        self.enqueue(dato)

    def buscar_elemento_recursivo(self, nodo, valor):
        if nodo is None:
            return False
        
        if nodo.dato == valor:
            return True
            
        return self.buscar_elemento_recursivo(nodo.siguiente, valor)

    def mostrar(self):
        actual = self.frente
        print("Cola: ", end="")
        while actual is not None:
            print(f"{actual.dato} ", end="")
            actual = actual.siguiente
        print()

if __name__ == "__main__":
    c = Cola()
    c.enqueue(10)
    c.enqueue(20)
    c.enqueue(30)

    print("Cola original:")
    c.mostrar()

    c.invertir_recursivo()
    print("Cola invertida recursivamente:")
    c.mostrar()

    print(f"Buscar 20 (recursivo): {c.buscar_elemento_recursivo(c.frente, 20)}")
    print(f"Buscar 50 (recursivo): {c.buscar_elemento_recursivo(c.frente, 50)}")
